engine/src/main/java/org/terasology/math/Diamond3iIterable.java

```diff
@@ -0,0 +1,162 @@
+        Preconditions.checkArgument(startDistance >= 0, "maxDistance must be >= 0");
+
+        this.origin.set(origin);
+        this.endDistance = maxDistance;
+        this.startDistance = startDistance;
+    }
+
+    /**
+     * solid diamond shape from the center to a radius
+     * @param origin center of diamond iterator
+     * @param radius distance to iterate out to
+     */
+    public static Diamond3iIterable.Builder region(Vector3ic origin, int radius) {
+        return new Diamond3iIterable.Builder(origin, radius);
+    }
+
+    /**
+     * hollow shape with an inner and outer distance to iterate over
+     * @param origin  center of diamond iterator
+     * @param start distance to start iterating from
+     * @param end distance to iterate out to
+     */
+    public static Diamond3iIterable.Builder hollow(Vector3ic origin, int start, int end) {
+        return new Diamond3iIterable.Builder(origin, end).start(start);
+    }
+
+    /**
+     * iterates in a shell that is 1 block wide
+     * @param origin center of shell
+     * @param radius distance for 1 block wide shell
+     */
+    public static Diamond3iIterable.Builder shell(Vector3ic origin, int radius) {
+        return new Diamond3iIterable.Builder(origin, radius).start(radius - 1);
+    }
+
+    @Override
+    public Iterator<Vector3ic> iterator() {
+        Vector3i pos = new Vector3i();
+        final int[] level = {this.startDistance};
+        final int[] offset = {-this.startDistance,0,0};
+
+        return new Iterator<Vector3ic>() {
+            @Override
+            public boolean hasNext() {
+                return level[0] < endDistance;
+            }
+
+            @Override
+            public Vector3ic next() {
+                pos.set(origin.x + offset[0], origin.y + offset[1], origin.z + offset[2]);
+                if (offset[2] < 0) {
+                    offset[2] *= -1;
+                } else if (offset[1] < 0) {
+                    offset[1] *= -1;
+                    offset[2] = -(level[0] - TeraMath.fastAbs(offset[0]) - TeraMath.fastAbs(offset[1]));
+                } else {
+                    offset[1] = -offset[1] + 1;
+                    if (offset[1] > 0) {
+                        if (++offset[0] <= level[0]) {
+                            offset[1] = TeraMath.fastAbs(offset[0]) - level[0];
+                            offset[2] = 0;
+                        } else {
+                            level[0]++;
+                            offset[0] = -level[0];
+                            offset[1] = 0;
+                            offset[2] = 0;
+                        }
+                    } else {
+                        offset[2] = -(level[0] - TeraMath.fastAbs(offset[0]) - TeraMath.fastAbs(offset[1]));
+                    }
+                }
+                return pos;
+            }
+        };
+    }
+
+    public static final class Builder {
```

Comment:
I'd prefer the inner class on top, but I'm not sure if we have style guidelines for that.